File: evals/behavioral/tasks/typescript/fixtures/conn-protocol.good.ts
// Typestate: each protocol state is its own type, so an operation only exists
// in the state where it's legal. `closedConn().send(..)` does not type-check
// (send only exists on an established connection).
export interface ClosedConn {
  listen(): ListeningConn;
}

export interface ListeningConn {
  accept(): EstablishedConn;
}

export interface EstablishedConn {
  send(data: string): EstablishedConn;
}

export function closedConn(): ClosedConn {
  function established(): EstablishedConn {
    return { send: (_data) => established() };
  }
  return { listen: () => ({ accept: () => established() }) };
}

File: evals/behavioral/tasks/python/fixtures/conn-protocol.good.py
# Typestate: each protocol state is its own class, so an operation only exists
# in the state where it's legal. `closed_conn().send(..)` is a pyright error
# (send only exists on an established connection).
class EstablishedConn:
    def send(self, data: str) -> "EstablishedConn":
        return self


class ListeningConn:
    def accept(self) -> EstablishedConn:
        return EstablishedConn()


class ClosedConn:
    def listen(self) -> ListeningConn:
        return ListeningConn()


def closed_conn() -> ClosedConn:
    return ClosedConn()
